### PR TITLE
Chapter/distributed computing completed

### PR DESCRIPTION
**Pull request by Ozzywap**
Distributed section complete. Added a parallel refresher to recap parallel computing section before diving into distributed computing as per the slides.

#7

### CODE REVIEW

File: src/chapter5/distributed-computing.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+## Advantages of distributed computing
+
+There are number of benefits to distributed computing in particular it addresses some shortcomings of shared memory architecture.
+
+- No contention for shared memory since each machine has its own memory. Compare this to shared memory architecture where all the cpu's are sharing the same memory.
+- Highly scalable as we can add more machines and are not limited by RAM.
+- Effectively resulting in being able to handle large-scale problems
+
+The benefits above do not come without some drawbacks including memory overhead.
```
Comment (Ozzywap) [inline src/chapter5/distributed-computing.md:32]: network overhead
```suggestion
The benefits above do not come without some drawbacks including memory overhead.
```

File: src/chapter5/distributed-computing.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+- Effectively resulting in being able to handle large-scale problems
+
+The benefits above do not come without some drawbacks including network overhead.
+
+## Disadvantages of distributed computing
+
+- Network overload. Network can be overloaded by:
+  - Multiple small messages
+  - Very large data throughput
+  - Multiple all-to-all messages (N^2 growth of messages)
```
Comment (oraqlle) [inline src/chapter5/distributed-computing.md:39]: You can use [mdBook's MathJax notation](https://rust-lang.github.io/mdBook/format/mathjax.html) for math stuff

File: src/chapter5/message-passing.md [hunk 3]
```diff
@@ -1 +1,240 @@
 # Message Passing
+
+## Primary MPI Routines
+
+``` C
+int MPI_Init(int * argc, char ** argv);
+// initializes the MPI environment.
+//Argc argv are the parameters come
```
Comment (oraqlle) [inline src/chapter5/message-passing.md:8]: Just make sure the C code style is consistent.

File: src/chapter5/message-passing.md [hunk 4]
```diff
@@ -1 +1,240 @@
...
+
+```C
+int MPI_Send(void * buf, int count, MPI_Datatype type, int dest, int tag, MPI_Comm comm);
+```
+
+Sends a message from the calling process to another process
+
+INPUT PARAMETERS
+
+- buf
```
Comment (oraqlle) [inline src/chapter5/message-passing.md:39]: Maybe put parameters in backticks eg. `buf`

File: src/chapter5/message-passing.md [hunk 5]
```diff
@@ -1 +1,240 @@
...
+
+### Primary MPI Routines closing
+
+In a header file you will find
+
+``` C
+int MPI_Finalize(void);
+```
+
+To call in your C/C++ program
```
Comment (oraqlle) [inline src/chapter5/message-passing.md:93]: Pedantic not but 'C/C++' should be 'C or C++'. The notation 'C/C++' often gets confused as if the language is 'C/C++' or the usual or meaning that / is used for.

File: src/chapter5/message-passing.md [hunk 6]
```diff
@@ -1 +1,240 @@
...
+
+Use man pages to find out more about each routine
+
+When sending a Process it packs up all of its necessary data into a buffer for the receiving process. These buffers are often referred to as envelopes since the data is being packed into a single message before transmission (similar to how letters are packed into envelopes before transmission to the post office)
+
+## Elementary MPI Data types
+
+MPI_Send and MPI_Recv utilize MPI Datatypes as a means to specify the structure of a message at a higher level. The data types defined in the table below are simple in nature and for custom data structures you will have to define the structure.
+
+![MPI datatypes](../imgs/mpi_datatypes.png)
```
Comment (oraqlle) [inline src/chapter5/message-passing.md:126]: Maybe create your own table in Markdown. [Markdown table generator](https://www.tablesgenerator.com/markdown_tables)

File: src/chapter5/message-passing.md [hunk 7]
```diff
@@ -1 +1,240 @@
...
+
+<https://docs.massive.org.au/M3/slurm/mpi-jobs.html>
+
+- ntasks                           Controls the number of tasks to be created for the job
+- ntasks-per-node           Controls the maximum number of tasks per allocated node
+- cpus-per-task               Controls the number of CPUs allocated per task
+
+## Measuring performance
+
+- htop to check the CPU usage. You need to run this command while the process is running
```
Comment (oraqlle) [inline src/chapter5/message-passing.md:222]: Should be `htop` to indicate command

File: src/chapter5/challenges.md [hunk 8]
```diff
@@ -1 +1,50 @@
 # Challenges
+
```
Comment (oraqlle) [inline src/chapter5/challenges.md:2]: Try and come up with new challenges. The old repo is not going to be used anymore so they're are not relevant.

**Review by oraqlle — CHANGES_REQUESTED** [review]
Looks good. Some small quality and consistency changes are needed. Try and come up with new challenges and remember to use the [challenge repo](https://github.com/MonashDeepNeuron/HPC-Training-Challenges) to add blueprint/template files and lay out the instructions as this will be where they will save there work. With regards to the Message Passing, try and discuss it from a high level first (which you started on in the 'distrub. comp.' section and then go into what OpenMPI is and how to use it.

File: src/chapter5/message-passing.md (hunk 3 above)
Comment (Ozzywap) [reply src/chapter5/message-passing.md:8]: Can I get you to elaborate on what you mean by this?

**Review by oraqlle — APPROVED** [review]
Looks good

**Pull request merged**